include/ginkgo/core/preconditioner/isai.hpp

```diff
@@ -0,0 +1,278 @@
+    using index_type = IndexType;
+    using Csr = matrix::Csr<ValueType, IndexType>;
+
+    /**
+     * Returns the approximate inverse of L.
+     *
+     * @returns the approximate inverse of L
+     */
+    std::shared_ptr<const Csr> get_approx_inverse_l() const { return l_inv_; }
+
+    /**
+     * Returns the approximate inverse of U.
+     *
+     * @returns the approximate inverse of U
+     */
+    std::shared_ptr<const Csr> get_approx_inverse_u() const { return u_inv_; }
+
+    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
+    {
+        /**
+         * When set to true, the Isai preconditioner expects a lower triangular
+         * factor (only, so no U factor), and will only generate the
+         * approximate inverse for that L factor.
+         *
+         * @note: Setting both exclusive_factor_l and exclusive_factor_u
+         *        results in an exception being thrown when generating the
+         *        preconditioner.
+         */
+        bool GKO_FACTORY_PARAMETER(exclusive_factor_l, false);
+
+        /**
+         * When set to true, the Isai preconditioner expects an upper triangular
+         * factor (only, so no L factor), and will only generate the
+         * approximate inverse for that U factor.
+         *
+         * @note: Setting both exclusive_factor_l and exclusive_factor_u
+         *        results in an exception being thrown when generating the
+         *        preconditioner.
+         */
+        bool GKO_FACTORY_PARAMETER(exclusive_factor_u, false);
+
+        /**
+         * @brief Optimization parameter that skips the sorting if the input
+         *        matrix/matrices (required by the algorithm).
+         *
+         * The algorithm to create the approximate inverses requires the
+         * input matrix/matrices to be sorted. If they are, this parameter
+         * can be set to `true` to skip the sorting for better performance.
+         */
+        bool GKO_FACTORY_PARAMETER(skip_sorting, false);
+    };
+
+    GKO_ENABLE_LIN_OP_FACTORY(Isai, parameters, Factory);
+    GKO_ENABLE_BUILD_METHOD(Factory);
+
+protected:
+    explicit Isai(std::shared_ptr<const Executor> exec)
+        : EnableLinOp<Isai>(std::move(exec))
+    {}
+
+    /**
+     * Creates an Isai preconditioner from a matrix using an Isai::Factory.
+     *
+     * @param factory  the factory to use to create the preconditoner
+     * @param factors  Composition<ValueType> of a lower triangular and an
+     *                 upper triangular matrix (L and U)
+     */
+    explicit Isai(const Factory *factory, std::shared_ptr<const LinOp> factors)
+        : EnableLinOp<Isai>(factory->get_executor(), factors->get_size()),
+          parameters_{factory->get_parameters()}
+    {
+        if (parameters_.exclusive_factor_l && parameters_.exclusive_factor_u) {
+            GKO_NOT_SUPPORTED(parameters_);
+        }
+        auto comp = dynamic_cast<const Composition<ValueType> *>(factors.get());
```

Comment:
I don't see any advantage for using `dynamic_pointer_cast` instead of `dynamic_cast` here.
I know that the object is at the very least valid until the end of the function (because the argument is a `std::shared_ptr`), and I don't need to store it anywhere or take ownership.
Using `dynamic_pointer_cast` has a bit more overhead (although that is neglectable here) and does not provide better readability than `dynamic_cast` in my opinion.